Bluetooth: controller: 255 byte AD payload support

cvinayak opened the pull request:
Add 255 byte AD payload support.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

~This is work in progress, todo:~
- ~Dynamic allocation of PDU buffer from pool.~
- ~mfifo based return path from LLL for stale/previous buffer~
- ~Wait for free buffer availability using k_sem_give/k_sem_take~

cvinayak committed:
Bluetooth: controller: 255 byte AD payload support

Add 255 byte AD payload support.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -71,6 +72,22 @@ static inline bool isr_rx_ci_tgta_check(struct lll_adv *lll,
+#endif
+
+#define PDU_MEM_SIZE  MROUND(PDU_AC_LL_HEADER_SIZE + PAYLOAD_SIZE_MAX)
+#define PDU_POOL_SIZE (PDU_MEM_SIZE * 7)
```
7 will be replaced with a Kconfig

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1530,11 +1541,18 @@ void ull_adv_done(struct node_rx_event_done *done)
 #endif /* CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL && !CONFIG_BT_CTLR_ADV_EXT */
 
+	for (handle = 0U; handle < BT_CTLR_ADV_SET; handle++) {
+		lll_adv_data_init(&ll_adv[handle].lll.adv_data);
```
primary channel AD Data to be restricted to 37 bytes max. Add new implementation specific to primary channel PDU allocation.

Thalley replied:
Is this a note or a TODO? Should it be added as a TODO in the code?

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1530,11 +1541,18 @@ void ull_adv_done(struct node_rx_event_done *done)
 
+	for (handle = 0U; handle < BT_CTLR_ADV_SET; handle++) {
+		lll_adv_data_init(&ll_adv[handle].lll.adv_data);
+		lll_adv_data_init(&ll_adv[handle].lll.scan_rsp);
```
legacy advertising scan response to be restricted to 37 bytes. Should 251 bytes buffer be used here for simplicity?

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -519,6 +519,12 @@ void ll_reset(void)
 
+#if defined(CONFIG_BT_BROADCASTER)
+	/* Init after adv state reset */
+	err = ull_adv_reset_init();
```
`ull_adv_reset` is called above, why is this call to `ull_adv_reset_init` necessary here?

cvinayak replied:
The order of reset procedure is:
1. Reset ULL context, like scheduling, abort LLL events etc.
2. Reset LLL context, like event members etc.
3. Reset ULL static members (which otherwise was cleared if on power up)
4. Init ULL context members

wopu-ot replied:
Can step 3. be done in step 1. (or vice versa)? If not, we should try to find better names, having ..._init, ..._reset, ..._reset_init and static functions called init_reset is confusing.

cvinayak replied:
I will some inline comments and find suitable names.

Thalley replied:
I guess the above comment still needs to be fixed w.r.t. renaming the confusing `init_reset` function.

wopu-ot replied:
I still don't understand why step 3 cannot be done as part of step 1. Can you please explain?

cvinayak replied:
step 3, cannot be done in step 1 because step 2 (call to `lll_adv_reset`) initializes the free AC PDU memory pool, which is needed in step 3.

You can suggest alternative names, current names reflect their purpose:
`ull_<role>_init` : Power up ULL initialization (Thread executable)
`lll_<role>_init`: Power up LLL initialization (Thread executable)
`ull_<role>_reset`: Warm reset ULL operation (Thread executable)
`lll_<role>_reset`: Warm reset LLL operation (LLL executable)
`ull_<role>_reset_init`: Initialization of ULL context after LLL warm reset operation (Thread executable)
`init_reset`: common code in `ull_<role>_init`, `ull_<role>_reset`, and `ull_<role>_reset_init` (Thread executable)

wopu-ot replied:
I think the _init/_reset functions are appropriately named. I considered some alternative names for `ull_<role>_reset_init`; the clearest I can come up with is `ull_<role>_finalize_reset` (as in "finalizing the ULL reset after LLL has been reset"). Unless the other reviewers prefer `ull_<role>_reset_init`, I propose to go for that.

cvinayak replied:
Do you agree with `ull_<role>_reset_finalize`? I would like to keep the verb as suffix.

wopu-ot replied:
That works for me.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -389,11 +394,6 @@ static int init_reset(void)
 	return 0;
 }
 
-static inline struct ll_adv_sync_set *sync_acquire(void)
-{
-	return mem_acquire(&adv_sync_free);
-}
-
```
It seems inconsistent to drop `sync_acquire` but keep `sync_release`. Please either keep both (which I would prefer) or drop both.

cvinayak replied:
I will revert.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -955,11 +961,6 @@ static int init_reset(void)
 	return 0;
 }
 
-static inline struct ll_adv_aux_set *aux_acquire(void)
-{
-	return mem_acquire(&adv_aux_free);
-}
-
```
It seems inconsistent to drop `aux_acquire` but keep `aux_release`. Please either keep both (which I would prefer) or drop both.

cvinayak replied:
I will revert.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c:
```diff
@@ -95,6 +112,75 @@ int lll_adv_reset(void)
+		return -ENOMEM;
+	}
+
+	pdu->pdu[0] = (void *)p;
```
> Two pdu in the lll_adv_pdu struct (e.g. pdu_ull and pdu_lll or whatever names makes the most sense)

`pdu->pdu[]` is ring buffer of 2 elements (pointers). 

> why we only init pdu[0] while we only reset pdu[1].

On power up, platform `_start`  instructions clear BSS section, i.e. pdu[0] and pdu[1] are NULL initialized; only pdu[0] need a valid memory allocated from pool. But, on HCI reset, pdu[0] needs to initialized to valid memory from pool, and pdu[1] be NULL initialized.

Thalley replied:
Is that documented anywhere? If not, perhaps it will be useful to have some documentation about the design and expected usage in the code.

cvinayak replied:
Will do.

wopu-ot replied:
The double-buffering seems to include yet another implementation of a FIFO. Could the code be simplified by using mfifo (or some other preexisting data structure)?

cvinayak replied:
double buffering is not a FIFO implementation, but replaces the latest AD data while maintaining context-safety when LLL is using the previous AD data during transmission.

cvinayak committed:
Bluetooth: controller: Release AD data on advertising set remove

Add implementation to release AD data PDU buffers on
advertising set remove.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Add semaphore to indicate free AD data buffers

Add semaphore implementation to indicate and wait for free
AD data buffers.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Minor move code closer to locality of reference

Minor change to move code closer to locality of reference.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Enable 251 byte AD data PDU

Enable the test procedure that requires greater than 31 byte
PDU when AD data and Sync Info structure be present in the
Advertising channel PDU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Add Max. AD data buffers Kconfig option

Add Kconfig option to configure maximum number of buffered
AD data PDUs across enabled advertising sets.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Use mfifo to return stale AD data PDU

Added implementation using mfifo to return stale AD data PDU
from LLL to Thread context.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -157,6 +157,16 @@ config BT_CTLR_ADV_SYNC_SET
 	  Maximum supported periodic advertising sets.
 
+config BT_CTLR_ADV_DATA_BUF_MAX
+	int "AD Data Max. Buffers"
```
I think it would be better to avoid abbreviations here, i.e. AD => Advertising Data and Max. => Maximum

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -157,6 +157,16 @@ config BT_CTLR_ADV_SYNC_SET
+	range 1 1 if !BT_CTLR_ADV_EXT
+	default 1
+	help
+	  Maximum no.of buffered AD Data payload across enabled advertising
```
Similar comment w.r.t. abbreviations here

Thalley reviewed (commented):
It generally looks good to me, but still missing some kind of documentation and description about the double buffer and the `lll_adv_data_` functions.

cvinayak commented:
> 
> 
> It generally looks good to me, but still missing some kind of documentation and description about the double buffer and the `lll_adv_data_` functions.

Yes, will add it. I found some bugs in this PR, will include as part of fixes here.

cvinayak committed:
Bluetooth: controller: Added comments related to AD data double buffering

Added inline comments explaining the implementation of AD
data PDU double buffering.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.h:
```diff
@@ -4,13 +4,23 @@
+ * the array element indexed by the member last.
+ * LLL in the ISR context, checks, traverses to the valid pointer indexed
+ * by the member first, such that the buffer is the latest commited by
+ * the thread context.
```
```suggestion
 * the thread context.
 */
```

andrzej-kaczmarek commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -48,6 +48,14 @@
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+#define PDU_AC_PAYLOAD_SIZE_MAX MAX(MIN(CONFIG_BT_CTLR_ADV_DATA_LEN_MAX, \
+					PDU_AC_EXT_PAYLOAD_SIZE_MAX), \
+				    PDU_AC_LEG_PAYLOAD_SIZE_MAX)
```
`BT_CTLR_ADV_DATA_LEN_MAX` is maximum length of AD, not PDU payload so even if e.g. `BT_CTLR_ADV_DATA_LEN_MAX 50` we need to support larger payload size to include common ext adv header. I think we should just assume `PDU_AC_EXT_PAYLOAD_SIZE_MAX` or find max overhead for common ext adv given our supported features in code.

cvinayak replied:
Max overhead of common ext adv header is 64 bytes, I have updated the PR now. And added a TODO comment for reducing the overhead based on features enabled in controller.

cvinayak committed:
Bluetooth: controller: Remove unused defined in pdu.h file

Remove used constant macro defined in pdu.h file.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix maximum AC PDU payload size calcuation

Fix maximum Advertising Channel PDU payload size calculation
by including the Common Extended Advertising Payload Format
overload alongwith the AD data maximum size supported.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

wopu-ot reviewed (changes requested):
The reset/init/reset_init/init_reset terminology is very confusing. Either the code should be simplified to avoid the subtle distinctions or more descriptive names are needed.
